tsdb/shard.go

```diff
@@ -1736,73 +1732,127 @@ func (fs *MeasurementFieldSet) IsEmpty() bool {
 	return len(fs.fields) == 0
 }
 
-func (fs *MeasurementFieldSet) Save() (err error) {
-	// current version
-	var v uint64
-	// Is the MeasurementFieldSet empty?
-	isEmpty := false
-	// marshaled MeasurementFieldSet
+type errorChannel chan<- error
 
-	b, err := func() ([]byte, error) {
-		fs.mu.Lock()
-		defer fs.mu.Unlock()
-		fs.memoryVersion += 1
-		v = fs.memoryVersion
-		// If no fields left, remove the fields index file
-		if len(fs.fields) == 0 {
-			isEmpty = true
-			if err := os.RemoveAll(fs.path); err != nil {
-				return nil, err
-			} else {
-				fs.writtenVersion = fs.memoryVersion
-				return nil, nil
-			}
-		}
-		return fs.marshalMeasurementFieldSetNoLock()
-	}()
+type MeasurementFieldSetWriter struct {
+	writeErrChannels chan errorChannel
+}
 
-	if err != nil {
-		return err
-	} else if isEmpty {
-		return nil
-	}
+// SetMeasurementFieldSetWriter - initialize the queue for write requests
+// and start the background write process
+func (fs *MeasurementFieldSet) SetMeasurementFieldSetWriter(queueLength int) {
+	fs.writer = &MeasurementFieldSetWriter{writeErrChannels: make(chan errorChannel, queueLength)}
+	go fs.saveWriter()
+}
 
-	// Write the new index to a temp file and rename when it's sync'd
-	// if it is still the most recent memoryVersion of the MeasurementFields
-	path := fs.path + "." + strconv.FormatUint(v, 10) + ".tmp"
+func (fs *MeasurementFieldSet) Save() error {
+	return fs.writer.RequestSave()
+}
 
-	fd, err := os.OpenFile(path, os.O_CREATE|os.O_RDWR|os.O_EXCL|os.O_SYNC, 0666)
-	if err != nil {
-		return err
-	}
-	defer os.RemoveAll(path)
+func (w *MeasurementFieldSetWriter) RequestSave() error {
+	done := make(chan error)
+	w.writeErrChannels <- done
+	return <-done
+}
 
-	if _, err := fd.Write(fieldsIndexMagicNumber); err != nil {
-		return err
+func (fs *MeasurementFieldSet) saveWriter() {
+	for {
+		// Block until someone modifies the MeasurementFieldSet and
+		// it needs to be written to disk.
+		errorChannels, err := fs.writeToFile(<-fs.writer.writeErrChannels)
+		for _, c := range errorChannels {
+			c <- err
+			close(c)
+		}
 	}
+}
 
-	if _, err := fd.Write(b); err != nil {
-		return err
-	}
+// writeToFile: Write the new index to a temp file and rename when it's sync'd
+func (fs *MeasurementFieldSet) writeToFile(first errorChannel) (errorChannels []errorChannel, err error) {
+	// Put the errorChannel on which we blocked into a slice to allow more invocations
+	// to share the return code from the file write
+	errorChannels = append(errorChannels, first)
```

Comment:
nitpick: I would write the error handling as:

defer func() {
  for _, c := range errorChannels {
    c <- err
  }
}()

I don't think the close() should be necessary, hopefully there are not multiple waiters on the channel.

Reply:
That's a nice style point.

Reply:
But you don't even need to return the error, in that case.  The `writeToFile` need not return anything.

Reply:
I like the `close` in case there is some magic clean-up, or GC preparation or whatever.  I don't see it as hurting anything, and being a good defensive practice.